### DISCUSSION TITLE
FM-радио [Обновлено 27.11.24]

### DISCUSSION THREAD

**Discussion by Chimildic** [Chimildic/goofy#35] [Избранное]
## Описание

Аддон позволяет получить треки с радиостанций в двух категориях: топ-40 и за день (только за сегодня). Можно накапливать треки в кэш на Google Диске.

## Установка

1. Создайте файл _addons_ или _radio_ и скопируйте в него следующий код

```js
// Описание https://github.com/Chimildic/goofy/discussions/35
const Radio = (function () {
  const URL_BASE = 'http://the-radio.ru/';
  return {
    getTracks: getTracks,
    getTopTracks: getTopTracks,
  };

  function getTracks(station, limit) {
    return parseTracklist(station, limit);
  }

  function getTopTracks(station, limit) {
    let stationResponse = fetchPageAsText(station)

    let variables = parseVariables(stationResponse.getContentText())
    let stationId = variables['MAIN_RADIO_ID']

    let setCookies = stationResponse.getAllHeaders()['Set-Cookie']
    let cookies = setCookies.map(cook => cook.split(',')).join(';')

    let cheerio = Cheerio.load(stationResponse.getContentText())
    let csrfToken = cheerio('#csrf_token').attr('value')

    let topTracks = CustomUrlFetchApp.fetch(`https://the-radio.ru/radio/${stationId}/top`, {
      method: 'post',
      contentType: 'application/x-www-form-urlencoded;',
      payload: `radio_id=${stationId}`,
      headers: {
        'cookie': cookies,
        'x-csrf-token': csrfToken,
      },
    })
    let tracknames = topTracks.map(track => parseNameTrack(track.title))
    return Search.multisearchTracks(Selector.sliceFirst(tracknames, limit))
  }

  function parseTracklist(station, limit) {
    let htmlContent = fetchPageAsText(station).getContentText()
    let variables = parseVariables(htmlContent)
    let tracknames = variables['MAIN_TRACKLIST_OF_RADIO'].map(track => parseNameTrack(track.title))
    return Search.multisearchTracks(Selector.sliceFirst(tracknames, limit))
  }

  function fetchPageAsText(station) {
    return CustomUrlFetchApp.fetch(`${URL_BASE}playlist/${station}`)
  }

  function parseVariables(htmlText) {
    let scriptData = ''
    for (let match of htmlText.matchAll(/<script type=\"text\/javascript\">(.+)<\/script>/g)) {
      if (match && match[1].trim().startsWith('var')) {
        scriptData = match[1]
        break;
      }
    }

    let variables = {}
    let match
    let variableRegex = /var\s+(\w+)\s*=\s*(.*?);/g;
    while (match = variableRegex.exec(scriptData)) {
      try {
        variables[match[1]] = JSON.parse(match[2])
      } catch {
        variables[match[1]] = eval(match[2])
      }
    }
    return variables
  }

  function parseNameTrack(item) {
    let [artist, track] = item.split(' - ')
    if (track) {
      track = track.split(/( x | feat*\w+| vs*\W+|,|\/|[(]| \+ id)/gi)[0]
      return `${artist} ${track}`.formatName()
    } else {
      return item.formatName()
    }
  }
})();
```

## Простой пример

Данные берутся с сайта [the-radio](http://the-radio.ru/). Найдите на нем желаемую станцию и скопируйте ее название из адресной строки. Например, для ссылки: `http://the-radio.ru/radio/relax-r201` названием будет `relax-r201`

```js
  function updateStation() {
    let tracks = Radio.getTracks('relax-r201')

    Filter.removeTracks(tracks, RecentTracks.get(2500))

    Playlist.saveWithReplace({
      name: 'Релакс',
      tracks: tracks
    })
  }
```

> Предыдущий вариант с сайтом _top-radio_ больше не работает.

## Комплексный пример

Парсинг и поиск по одному треку в совокупности с большим количеством радиостанций быстро приводят к лимиту выполнения (6 минут).  [Clerk](https://chimildic.github.io/goofy/#/reference/clerk) поможет распределить N станций на один триггер, обновив за день все из них.

| Константы       	| Описание                                                                                                                                                                                              	|
|-----------------	|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------	|
| FIRST_HOURS     	| Первый час, с которого начинается обновление. Источник радио дает историю только за день. Поэтому начинать в час ночи не имеет смысла, история слишком маленькая. Утром или в середине дня уже лучше. 	|
| STATIONS        	| Перечень id от желаемых радиостанций. Имя будет использоваться в качестве названия плейлиста.                                                                                                         	|
| STATION_PER_RUN 	| Сколько станций обновить за один запуск по расписанию. Каждый запуск обновит следующие N станций. Если пройден весь список, ничего не произойдет.                                                     	|
| Общие источники 	| Всегда нужны общие источники вроде истории прослушиваний или лайков. Чтобы не брать их на каждую станцию отдельно, выносим в общую область.                                                           	|

- Функции `updateAllStations` ставим триггер на каждый час
- Функции `run` и `onQueue` системные, не трогаем
- Функция `updateStation` - то, что будет происходит с каждой отдельной станцией при запуске по расписанию. В этом примере это просто обновление плейлиста. Можете заменить на накопление кэша. Обратите внимание как используется общий источник `recentTracks`.

```js
function updateAllStations() {
  const FIRST_HOURS = 6
  const STATION_PER_RUN = 1
  const STATIONS = [
    { id: 'record-dance-r1721', name: 'Record DanceCore' },
    { id: 'uplifting-r2822', name: 'Record Uplifting' },
    { id: 'europa-plus-r363', name: 'Europa Plus' },
    { id: 'dorozhnoe-r157', name: 'Дорожное' },
    { id: 'relax-r201', name: 'Relax' },
    { id: 'nrg-r200', name: 'Energy' },
    { id: 'hit-fm-r195', name: 'Hit FM' },
    { id: 'maximum-r962', name: 'Maximum' },
    { id: 'rock-fm-r1903', name: 'Rock FM' },
    { id: 'studio-21-r2335', name: 'Studio 21' },
  ]

  // Общие источники
  let recentTracks = RecentTracks.get(2500)
  // ...
  // ...

  run()

  function run() {
    globalThis.onQueue = onQueue
    for (let i = FIRST_HOURS; i < 24; i++) {
      let timeStr = `${i}`.padStart(2, '0') + ':00'
      let func = new Function('return ' + `function onHour${i}() { onQueue(${i}) }`)()
      Clerk.runOnceAfter(timeStr, func)
    }
    Admin.printInfo('Количество запросов', CustomUrlFetchApp.getCountRequest())
  }

  function onQueue(hours) {
    let index = hours - FIRST_HOURS
    if (index < 0) {
      Admin.printError('Индекс меньше нуля')
      return
    }
    let stations = STATIONS.slice(index, index + STATION_PER_RUN)
    stations.forEach(s => updateStation(s.id, s.name))
  }

  function updateStation(id, name) {
    let tracks = Radio.getTracks(id)

    Filter.removeTracks(tracks, recentTracks)

    Playlist.saveWithReplace({
      name: name,
      tracks: tracks
    })
  }
}
```

**Comment by amd64fox** [comment]
Спасибо огромное за новую, дельную фишку... Все работает 😁

**Comment by tupizm** [comment]
Я только не совсем понял, топ за день меняется каждый день? И на основе чего этот топ? Я просто не пользовался этим сервисом

**Reply by tupizm** [reply]
Кажется понял. В течение дня играют песни на радио - создается плейлист дня, так сказать. После этого пользователи слушают эти песни через вк или ютуб, и на основе прсолушанных составляется топ дня. Соответсвенно каждый день песни будут уникальными, в каком-то смысле. я правильно понял? поправьте, пожалуйста, если вдруг нет

**Reply by Chimildic** [reply]
Треки за день - просто что играло на радио. Топ - наиболее часто игравшие. 
Когда топ обновляется не написано. Может быть за неделю. На практике проверить нужно.

**Reply by tupizm** [reply]
Блин, так это можно получается свое радио запилить в споти на основе этих треков за день. Это же просто бомба!

**Reply by efremych** [reply]
Другим сервисом пользуюсь, тоже обращал внимание на Топ, но не придавал ему значения. Да, на практике нужно следить, врядли ежедневно обновляется, но всё равно круто.

А, не заметил что кроме топа есть еще и треки за день - то что надо :)

**Comment by amd64fox** [comment]
Кстати после добавления фишки с радио, в каждом скрипте появилось предупреждение вида :

`3 февр. 2021 г., 18:03:57	Предупреждение	@babel/polyfill is loaded more than once on this page. This is probably not desirable/intended and may have consequences if different versions of the polyfills are applied sequentially. If you do need to load the polyfill more than once, use @babel/polyfill/noConflict instead to bypass the warning.`

Это нормально ?

**Reply by Chimildic** [reply]
У меня такого нет. Последнюю версию `Cheerio` выбрали?

**Reply by amd64fox** [reply]
Хм... странно выбирал последнюю при установке , сейчас проверил стоит 2-ая, исправил на 12-ую вручную

**Reply by amd64fox** [reply]
Да, из-за версии, исчезло предупреждение.

**Comment by amd64fox** [comment]
Еще заметил что есть повторы в списках, возможно нужно их убирать сразу или нет, не критично

**Reply by Chimildic** [reply]
В топе не должно быть по идеи. В треках дня логично, что есть, ведь эфир с радио.

**Comment by Bedmit** [comment]
Что-то я похоже упустил из виду. 
На скрипт 

```
function` Test() {
  let Test = Radio.getTracks('maximum');

  Playlist.saveAsNew({
  name: 'Test',
  tracks: Test,
});
}
```
Выдает ошибку 
```
SyntaxError: Identifier 'Radio' has already been declared
(функция без названия)	@ Radio.gs:1
```

Скрипт с выложенным кодом создал. Библиотеку добавил.

**Reply by Chimildic** [reply]
Имя `Radio` уже существует. Возможно, одна из личных функций так названа. Поэтому конфликт с модулем радио.

**Reply by Bedmit** [reply]
Точно. Уже использую это название
`function Radio(playlistUrl)`
Сейчас поменяю

**Reply by Bedmit** [reply]
Все, теперь все отлично работает. Спасибо!

**Comment by efremych** [comment]
А как вот такие вещи пресекать?
Чтобы название артиста было без выделенного текста?

3 февр. 2021 г., 22:52:39	Информация	Spotify по типу track не нашел: juno reactor **_vs don davis_** navras
3 февр. 2021 г., 22:52:39	Информация	Spotify по типу track не нашел: unkle _**featuring ian brown**_ be there

Так понимаю где-то здесь?
Но если экспериментировать много запросов уходит :(

`let artist = cheerio(node).text().split(/,|&|x|feat|/)[0];`

И "The" перед названием артиста. 
Добавил сюда, но оказалось, негативно влияет на артиста **Amaranthe** (в такой вариации с пробелом ' ' лучше, но всё же, может как-то правильнее можно)
   ```js
 function parseNameTrack(str) {
        return str
            .formatName()
            .replace(/( x )/g, ' ')
            .replace(/(the )/g, '')
            .replace(/( feat )/g, ' ');
    }
```

**Reply by efremych** [reply]
Всё, получилось, вот так поиск находит такие треки.
Понаблюдаем...
```js
    function parseNameTrack(str) {
        return str
            .formatName()
            .replace(/( x )/g, ' ')
            .replace(/^(the )|( vs )|( featuring )/g, ' ')
            .replace(/( feat )/g, ' ');
    }
```

**Reply by Chimildic** [reply]
Обновил `Radio`. Попробуйте заменить весь код. 
Кроме варианта с `the`. Его не понял. Можно название станции с этим треком?

**Reply by efremych** [reply]
Завтра посмотрю, комп выключил.
Радио ‘ultra’. 
Плейлист за 1 февраля, как в примере. Артист “The Fever 333”. Трек «One of Us”.

Кстати, возникла мысль, актуальная как для этого радио так и Истории Ласт Фм и сторонних платформ в принципе. В плане оптимизации количества запросов. 

Можно было бы наряду с атрибутами трека, который прошёл уже однажды процедуру поиска в Спотифае, сохранять в кэше также и поисковую строку «Артист-трек». Чтобы в следующий раз, когда надо будет опять найти этот трек, не лезть в Спотифай, а посмотреть в кэше, что соответствует данной поисковой строке

**Reply by efremych** [reply]
Обновил, да, теперь осталось только "the ". Остальное находит.

4 февр. 2021 г., 08:27:58	Информация	Spotify по типу track не нашел: the fever 333 one of us

P.S. Насчет, мысли с поиском+кэшем, наверно, не вариант, так как лучший результат поиска может меняться на глазах и да и трек может стать недоступным. Разве что, в пределах одних суток не дублировать обращения или "на усмотрение пользователя" оставлять как искать треки.

**Reply by Chimildic** [reply]
У группы искаженное название. При поиске в интернете, в одних местах с `the` пишется, в других без. И сам Spotify молодец, пара лишних букв ломает поиск. Вопрос в том, насколько повлияет удаление этой частицы на другие треки.

**Reply by efremych** [reply]
готов у себя рискнуть локально :) по идее поиску должно быть проще если удалить частицу, чем наоборот...
Но нужен совет, как удалять "the " в начале названия исполнителя.

**Reply by Chimildic** [reply]
Попробуйте так, не проверял
```js
let artist = item.artist.split(re)[0].replace(/^the /i, '');
```
Для удобства, можно использовать онлайн редакторы регулярных выражений, например [regex101](https://regex101.com/)

**Reply by efremych** [reply]
> let artist = item.artist.split(re)[0].replace(/^the /i, '');
Да, спасибо, подходит

> например regex101
Сколько еще всего интересного, оказывается

**Comment by Bedmit** [comment]
Отличная штука получилась!!!
Мне как всегда не хватает его поддержки аддоном запуска с телефона. Чтобы был один скрипт, а ссылку на радиостанцию и его запуск делать ситуативно. 
И еще мне не хватает таблеток от жадности )

**Reply by Chimildic** [reply]
Добавил возможность создать форму с предопределенными значениями. В том же [посте](https://github.com/Chimildic/goofy/discussions/9).

**Comment by CTpaTer** [comment]
let topTracks = Radio.getTopTracks('energy');
А как правильно прописать получение плейлиста с двух и более радиостанций?

**Reply by amd64fox** [reply]
```javascript
let topTracks = Radio.getTopTracks('energy');
let topTracks2 = Radio.getTopTracks('evropa-plyus');  

Combiner.push(topTracks, topTracks2); // если нужно добавить все в один массив.
```

**Reply by CTpaTer** [reply]
Спасибо!
Сколько раз сам себе говорил: "Читай документацию!" :-)

**Comment by periscop-mail** [comment]
Перестал работать скрипт. Как пофиксить?

**Reply by Chimildic** [reply]
Сайт _top-radio_ больше не позволяет парсить треки. Изменил код под другой источник. Прочитайте первый пост в [данной теме](https://github.com/Chimildic/goofy/discussions/35#discussion-2671327).

**Comment by Ciceronn** [comment]
Привет,помогите пожалуйста заполнить скрипт для радио максимум,url base я заполнил,но не работает почему-то.Что я сделал не правильно?Сорян заранее за возможно глупый вопрос,я ничерта не понимаю в этих скриптах

**Reply by Chimildic** [reply]
Вам не нужно изменять `url base`. Просто целиком скопируйте код к себе без изменений. Все пошагово описано в первом посте этой темы. В случае [радио максимум](http://the-radio.ru/radio/maximum-r962), используйте функцию так
```js
let topTracks = Radio.getTopTracks('maximum-r962');
let tracks = Radio.getTracks('maximum-r962');
```

**Reply by Ciceronn** [reply]
Простите за глупость,но куда вставлять эти 2 строки?Что значит "используйте функцию так"?Как именно?

Создал новый скрипт,во вторую строку вставил код приведенный вами.А куда вставлять строки let...?

Библиотека создана,чирио 14 добавлен

**Reply by Chimildic** [reply]
Код из первого поста вам нужно скопировать в корень файла. На уровне других функций. Для удобства можно создать новый файл.
Теперь пишем отдельную функцию, которая будет использовать скопированный код. К примеру:
```js
function updateRadioMaximum() {
  let tracks = Radio.getTracks('maximum-r962');
  Playlist.saveWithReplace({
    name: 'Радио Максимум',
    tracks: tracks,
  })
}
```

**Reply by Ciceronn** [reply]
Спасибо вам большое,все заработало!

**Reply by Mordva68** [reply]
А как сделать такую же функцию на toptracks? поменял tracks на toptracks, но появляется ошибка.... Что я делаю не так?

function updateRadioMaximum() {
  let tracks = Radio.getTracks('maximum-r962');
  Playlist.saveWithReplace({
    name: 'Радио Максимум',
    tracks: tracks,
  })
}

**Reply by Chimildic** [reply]
@Mordva68 
```js
function updateRadioMaximum() {
  let tracks = Radio.getTopTracks('maximum-r962');
  Playlist.saveWithReplace({
    name: 'Радио Максимум',
    tracks: tracks,
  })
}
```

**Comment by kysskader** [comment]
Здравствуйте! нужна помощь! Создал скрипт, проверил - работает, удалил созданные плейлисты, заменил станции на свои, запускаю скрипт - а в ответ "18:14:32	Примечание	Выполнение начато
18:14:13	Информация	Количество запросов 0
18:14:35	Примечание	Выполнение завершено"
Как вернуть плейлисты? Почему не работает после добавления станций?

**Reply by Chimildic** [reply]
Не у всех станций есть история треков. Также проверьте id станции.

**Reply by kysskader** [reply]
Осталась часть станций из примера. И они не хотят больше работать...

**Comment by andrey133mhz** [comment]
Всем, привет!
Кто-нибудь сталкивался с такой проблемой, что скрипт выдаёт ошибку по созданию созданию плейлиста, видимо из-за ошибки по названию станции? Например, такой косяк со станцией D'n'B Classics (Радио Рекорд). История и топ треков в этой станции показывает. При этом забиваю в скрипт другую соседнюю станцию, и всё работает. На всякий случай скрин с ошибками.

![Снимок](https://github.com/Chimildic/goofy/assets/85558615/3b13ba7c-5dda-4ed5-b0f4-a2014e022f54)

И ещё вопрос, синтаксически скрипт воспринимает слово "rmx" как "remix", а то часть треков в спотике добавляются либо оригинал композиции, либо он её не находит. И если не воспринимает, есть ли какие-то решения?
Заранее спасибо!

**Reply by Chimildic** [reply]
Приложите код станции с ошибкой

Скрипт лишь отправляет запрос к спотифай. Можно добавить подмены rmx на remix, но не факт что поможет

**Reply by andrey133mhz** [reply]
```
  function updateStationPirateStation() {
    let tracks = Radio.getTracks('drumnbass-hits-r869')
    Filter.removeTracks(tracks, RecentTracks.get(2500))
    Playlist.saveWithReplace({
      name: 'Pirate Station',
      tracks: tracks
    })
}
```

**Reply by andrey133mhz** [reply]
и например, когда я вставляю другое название "pirate-station-r359", то всё работает

**Reply by Chimildic** [reply]
Из-за символа кавычек в названии станции, который кодируется через специальный код, парсер ломается.

В коде аддона найдите строку, которая начинается с `let variableRegex` и замените на эту `let variableRegex = /var\s+(\w+)\s*=\s*(.*?);\s+/g;`.

**Reply by andrey133mhz** [reply]
Спасибо большое!!
Заработало
